docs: store magento2 credentials in the global config, for #5932, for #6380 (#6442)
ddev/ddev: docs/content/users/quickstart.md

    Normal details of a Composer build for Magento 2 are on the [Magento 2 site](https://experienceleague.adobe.com/docs/commerce-operations/installation-guide/composer.html). You must have a public and private key to install from Magento’s repository. When prompted for “username” and “password” in `composer create`, it’s asking for your public key as "username" and private key as "password".

    !!!tip "Store Adobe/Magento Composer credentials in the global DDEV config"

>         If you already have Composer installed on you system you can reuse the `auth.json` by making a symlink. See [In-Container Home Directory and Shell Configuration](extend/in-container-configuration.md):

Reviewer: ```suggestion
        If you have Composer installed on your workstation and have an `auth.json` you can reuse the `auth.json` by making a symlink. See [In-Container Home Directory and Shell Configuration](extend/in-container-configuration.md):
```

        ```
        mkdir -p ~/.ddev/homeadditions/.composer && ln -s ~/.composer/auth.json ~/.ddev/homeadditions/.composer/auth.json
        ```

>         Alternatively, you can install the Adobe/Magento Composer credentials in your global `~/.ddev/homeadditions/.composer/auth.json` and never have to find them again (see below):

Reviewer: ```suggestion
        Alternately, you can install the Adobe/Magento Composer credentials in your global `~/.ddev/homeadditions/.composer/auth.json` and never have to enter them again (see below):
```

    ```bash
    mkdir my-magento2-site && cd my-magento2-site
    ddev config --project-type=magento2 --docroot=pub --disable-settings-management \
    --upload-dirs=media

    ddev get ddev/ddev-elasticsearch
    ddev start

    # if the Magento auth is not found in the Composer global config, save it at the project level
    if ! ddev exec "composer config --list --global 2>/dev/null | grep -q repo.magento.com" 2>/dev/null; then
        ddev config --web-environment-add=COMPOSER_HOME="/var/www/html/.ddev/homeadditions/.composer"
        mkdir -p .ddev/homeadditions/.composer
        echo "*\n\!.gitignore" >.ddev/homeadditions/.composer/.gitignore
        ddev restart

>     fi

Author: I didn't test `echo "*\n\!.gitignore"` with bash 😅
It doesn't work as expected there, but it is fine with zsh, I will fix it.

>     fi

Author: > was it supposed to gitignore the auth.json?

Yes, it was supposed to ignore everything in that folder except `.gitignore`, but the resulting file in bash looked like this:

```
*\n\!.gitignore
```

instead of

```
*
!.gitignore
```

    ddev get ddev/ddev-elasticsearch
    ddev start

    # if the Magento auth is not found in the Composer config, save it at the project level

>     if ! ddev exec "composer config --list 2>/dev/null | grep -q repo.magento.com || composer config --list --global 2>/dev/null | grep -q repo.magento.com" 2>/dev/null; then

Reviewer: I might be wrong, but I don't think that `composer config --list --global` is relevant here. The difference is between *DDEV*'s global and local homeadditions, both of which land in the *user* directory inside the project, `~/.composer/auth.json` - So maybe my comment was naive in the first place. Anyway, nothing we do makes it into composer's global config.

Author: > So maybe my comment was naive in the first place.

Your suggestion was correct because someone could override `COMPOSER_HOME` in the DDEV's global config.

> both of which land in the user directory inside the project, `~/.composer/auth.json`

Not both, when you have `COMPOSER_HOME` set, it will use a different path.

> I don't think that `composer config --list --global` is relevant here.

It's relevant because it runs after `ddev start`, so `~/.ddev/homeadditions/.composer/auth.json` is in `~/.composer/auth.json` at this point.

It just needs some manual testing for a better understanding.

Author: I reverted changes for `composer config --list` as it only shows the config for `composer.json` in the current directory.

On the other hand, `composer config --list --global` shows the config you have in `~/.composer/auth.json` or `COMPOSER_HOME` inside the container.

Reviewer: I was confused, of course. `global` config in composer is (normally) ~/.composer, and regular is in the project composer.json.
